src/mlpack/methods/reinforcement_learning/environment/frozen_lake.hpp

```diff
@@ -0,0 +1,470 @@
+   * and a new random environment board is created.
+   *
+   * @return Initial state for each episode.
+   */
+  State InitialSample()
+  {
+    stepsPerformed = 0;
+    boardDescription = generateRandomBoard();
+    return State(height, width);
+  }
+
+  /**
+   * Initial state representation is the current position at (0, 0) 
+   * and a new random bord is created.
+   * 
+   * @param board Environment board of that the user want to initialize to.
+   * @param height Height of the environment board.
+   * @param width Width of the environment board.
+   * @return A state in which the height and width of the environment board are 
+   *    limited to height and width, respectively.
+   */
+  State InitialSample(arma::Mat<size_t> board, size_t height, size_t width)
+  {
+    stepsPerformed = 0;
+    this->height = height;
+    this->width = width;
+    boardDescription = board;
+    return State(height, width);
+  }
+
+  /**
+   * This function checks if the agent has reached the terminal state.
+   *
+   * @param state The desired state.
+   * @return true if state is a terminal state, otherwise false.
+   */
+  bool IsTerminal(const State& state) const
+  { 
+    if (maxSteps != 0 && stepsPerformed >= maxSteps)
+    {
+      Log::Info << "Episode terminated due to the maximum number of steps"
+          "being taken.\n";
+      return true;
+    }
+    else if (boardDescription(state.CurRow(), state.CurCol()) == tiles::Hole)
+    {
+      Log::Info << "Episode terminated due to agent falling in the hole.\n";
+      return true;
+    }
+    else if (boardDescription(state.CurRow(), state.CurCol()) == tiles::Goal)
+    {
+      Log::Info << "Episode terminated, agent reached the goal.\n";
+      return true;
+    }
+    return false;
+  }
+
+  //! Get the number of steps performed.
+  size_t StepsPerformed() const { return stepsPerformed; }
+
+  //! Get the maximum number of steps allowed.
+  size_t MaxSteps() const { return maxSteps; }
+
+  //! Set the maximum number of steps allowed.
+  size_t& MaxSteps() { return maxSteps; }
+
+  //! Get the board description.
+  // std::vector<std::vector<char>> Description() const {return boardDescription;}
+  arma::Mat<size_t> Description() const {return boardDescription;}
+
+  //! Set the board description.
+  // std::vector<std::vector<char>>& Description() {return boardDescription;}
+  arma::Mat<size_t>& Description() {return boardDescription;}
+
+  enum tiles
+  {
+    Start, Goal, Frozen, Hole
+  };
+
+ private:
+  /**
+   * Utility function that helps generate random environment board.
+   * 
+   * @return board, a 2D array of characters that hold the board description.
+   *    'S' is the Start Tile, the starting position of the agent. 0
+   *    'G' is the Goal Tile, the agent can walk into a goal tile to win. 1
+   *    'F' is a Frozen Tile, the agent can walk on it. 2
+   *    'H' is a Hole Tile, the agent can fall into it and the game ends. 3
+   */
+  arma::Mat<size_t> generateRandomBoard()
```

Comment:
```suggestion
  arma::Mat<size_t> GenerateRandomBoard()
```

We use upper-camel casing in mlpack.

Reply:
sure! Will change!